File: src/backend/yt_diffuser/usecases/web/project/layer_usecase.py
from injector import inject

from yt_diffuser.types.project import Project, ProjectLayer
from yt_diffuser.adapters.project.interface import IProjectLayerRepository

class ProjectLayerUseCase:
    """
    プロジェクトのレイヤーに関するユースケース
    """

    @inject
    def __init__(self, layer_repository: IProjectLayerRepository):
        self.layer_repository = layer_repository

    def is_exists(self, project_name: str, layer: ProjectLayer) -> bool:
        """
        レイヤーが存在するか確認する
        """
        return self.layer_repository.is_exists(project_name, layer)
    
    def add_layer(self, project_name:str, layer: ProjectLayer) -> None:
        """
        レイヤーを追加する
        """
        if self.is_exists(project_name, layer):
            raise Exception(f"Layer {layer.layer_id} already exists.")

        self.layer_repository.create(project_name, layer)
        return

    def upload_image(self, project_name:str, layer_id:str, file_name:str, source_path:str) -> None:
        """
        画像をアップロードする
        """
        self.layer_repository.save_image(project_name, layer_id, file_name, source_path, is_copy=False)
        return

File: src/frontend/src/composables/project/layerUseCase.js
import { toRef } from 'vue'
import { useAPI } from '@/adapters/api'
import { useProjectStore } from '@/stores/app/'
import { ProjectLayer } from '@/types/project'
import { ImageFile } from '@/types/image'
import { useProjectUseCase } from './projectUseCase'

/**
 * レイヤーのユースケースを返す
 *
 * @returns {ReturnType<typeof ProjectLayerUseCase>}
 */
export function useProjectLayerUseCase() {
  return ProjectLayerUseCase(useAPI(), useProjectStore(), useProjectUseCase())
}

/**
 * レイヤーのユースケース
 *
 * @param {ReturnType<typeof useAPI>} api
 * @param {ReturnType<typeof useProjectStore>} projectStore
 * @param {ReturnType<typeof useProjectUseCase>} projectUseCase
 * @returns
 */
export function ProjectLayerUseCase(api, projectStore, projectUseCase) {
  const project = toRef(projectStore, 'data')
  return {
    /**
     * プロジェクトストアの参照を返す
     */
    getRefs() {
      return {
        project,
        selectedLayer: toRef(projectStore, 'selectedLayer'),
        isOpen: toRef(projectStore, 'isOpen')
      }
    },
    /**
     * レイヤーを取得する
     */
    getLayer(layerId) {
      return project.value.layers.layers[layerId]
    },
    /**
     * レイヤー画像のパスを取得
     *
     * @param {String} layerId
     * @param {String} imageType 表示するイメージファイル (image, i2i, mask)
     * @return {String} 画像のパス 画像が存在しない場合は空文字を返す
     */
    getImageUrl(layerId, imageType) {
      const projectName = project.value.projectName
      const image = project.value.layers.layers[layerId][imageType]

      if (image) {
        return `output/project/${projectName}/layers/${layerId}/${image}`
      } else {
        return ''
      }
    },
    /**
     * レイヤーを選択する
     */
    selectLayer(layerId) {
      projectStore.selectLayer(layerId)
    },

    /**
     * レイヤーを追加する
     */
    async addLayer(layerName) {
      await api.post(`/api/project/${project.value.projectName}/layer`, { layerName })
      await projectUseCase.load(project.value.projectName)
    },

    /**
     * 選択中のレイヤーに画像をアップロードする
     *
     * @param {String} imageType 画像の種類 (image, i2i, mask)
     * @param {File} file 画像ファイル
     */
    async uploadImage(imageType, file) {
      const projectName = project.value.projectName
      const layerId = projectStore.selectedLayer
      await api.upload(`/api/project/${projectName}/layer/${layerId}/${imageType}`, { file })
      await projectUseCase.load(projectName)
    }
  }
}
